consensus/core/src/metrics.rs

```diff
@@ -77,24 +84,354 @@ const SIZE_BUCKETS: &[f64] = &[
     10_000_000.0,
 ]; // size in bytes
 
+const ERRORS_FROM_UNSIGNED_BLOCK_VERIFICATION: [&str; 6] = [
+    "WrongEpoch",
+    "UnexpectedGenesisBlock",
+    "InvalidAuthorityIndex",
+    "SerializationFailure",
+    "MalformedSignature",
+    "SignatureVerificationFailure",
+];
+
+const ERRORS_FROM_SIGNED_BLOCK_VERIFICATION: [&str; 11] = [
+    "TooManyAncestors",
+    "InsufficientParentStakes",
+    "InvalidAuthorityIndex",
+    "InvalidAncestorPosition",
+    "InvalidAncestorRound",
+    "InvalidGenesisAncestor",
+    "DuplicatedAncestorsAuthority",
+    "TransactionTooLarge",
+    "TooManyTransactions",
+    "TooManyTransactionBytes",
+    "InvalidTransaction",
+];
+
+const UNPROVABLE_ERRORS_FROM_COMMIT_SYNCER: [&str; 8] = [
+    "MalformedCommit",
+    "UnexpectedStartCommit",
+    "UnexpectedCommitSequence",
+    "NoCommitReceived",
+    "MalformedBlock",
+    "NotEnoughCommitVotes",
+    "UnexpectedNumberOfBlocksFetched",
+    "UnexpectedBlockForCommit",
+];
+
+const UNPROVABLE_ERRORS_FROM_SUBSCRIBER: [&str; 2] = ["MalformedBlock", "UnexpectedAuthority"];
+
+const PROVABLE_ERRORS_FROM_SUBSCRIBER: [&str; 2] = ["BlockRejected", "MalformedAncestorBlock"];
+
 pub(crate) struct Metrics {
     pub(crate) node_metrics: NodeMetrics,
     pub(crate) network_metrics: NetworkMetrics,
+    pub(crate) scoring_metrics: ValidatorScoreMetrics,
 }
 
-pub(crate) fn initialise_metrics(registry: Registry) -> Arc<Metrics> {
-    let node_metrics = NodeMetrics::new(&registry);
-    let network_metrics = NetworkMetrics::new(&registry);
+impl Metrics {
+    pub(crate) fn update_scoring_metrics_on_eviction(
+        &self,
+        validator: AuthorityIndex,
+        hostname: &str,
+        eviction_round: u32,
+        last_evicted_round: u32,
+        recent_refs_by_authority: &BTreeSet<BlockRef>,
+        threshold_clock_round: u32,
+    ) -> Option<StoredScoreMetricsU64> {
+        // Get the blocks rounds that were not evicted.
+        let mut cached_block_rounds = recent_refs_by_authority
+            .iter()
+            .map(|block| block.round)
+            .filter(|&round| round > eviction_round && round < threshold_clock_round)
+            .collect::<Vec<u32>>();
 
-    Arc::new(Metrics {
-        node_metrics,
-        network_metrics,
-    })
-}
+        // Update metrics according to the blocks from rounds still in cache.
+        let number_of_cached_blocks = cached_block_rounds.len();
+        cached_block_rounds.dedup();
+        let unique_cached_block_rounds = cached_block_rounds.len();
+        let cached_equivocations =
+            number_of_cached_blocks.saturating_sub(unique_cached_block_rounds) as u64;
+        let missing_blocks_in_cached_rounds = threshold_clock_round
+            .saturating_sub(eviction_round + unique_cached_block_rounds as u32 + 1)
+            as u64;
 
-#[cfg(test)]
-pub(crate) fn test_metrics() -> Arc<Metrics> {
-    initialise_metrics(Registry::new())
+        self.scoring_metrics.cached[validator.value()]
```

Comment:
We have `Index<AuthorityIndex> for Vec<T>` so there is no need to call `.value()`.